### PR TITLE
perf: weaken power in twiddles generation

### PR DESCRIPTION
**Pull request by Oppen**
Because we save all powers, we can extract the computation from the loop and turn it into successive products instead.
This improves twiddle generation from 5 to 23 times depending on config, and FFT itself by about 10%.
Additional improvement by replacing the batched inverse by simply computing the inverse of the root ahead of time.

Uses the benchmarks introduced in #373 

## Type of change

- [x] Optimization

## Checklist
- [x] This change is an Optimization
  - [x] Benchmarks added/run

Results (ran in MacBook Pro M1 2020 16GB with MacOS Ventura 13.3.1:
```
Ordered FFT/Sequential from NR radix2                                                                                                                                                              [17/1824]
                        time:   [676.77 ms 677.22 ms 677.82 ms]
                        thrpt:  [3.0940 Melem/s 3.0967 Melem/s 3.0988 Melem/s]
                 change:
                        time:   [-20.765% -20.376% -20.003%] (p = 0.00 < 0.05)
                        thrpt:  [+25.004% +25.590% +26.207%]
                        Performance has improved.

Ordered FFT/Sequential from RN radix2                                                                           
                        time:   [1.2457 s 1.2508 s 1.2559 s]                                                                                                                                                
                        thrpt:  [1.6699 Melem/s 1.6767 Melem/s 1.6835 Melem/s]
                 change:                                                                              
                        time:   [-17.362% -15.432% -13.770%] (p = 0.00 < 0.05)
                        thrpt:  [+15.969% +18.248% +21.009%]                  
                        Performance has improved.                                                     
                                                   
FFT twiddles generation/natural                                                                            
                        time:   [19.367 ms 19.426 ms 19.510 ms]
                        thrpt:  [53.746 Melem/s 53.978 Melem/s 54.143 Melem/s]
                 change:
                        time:   [-96.929% -96.916% -96.902%] (p = 0.00 < 0.05)
                        thrpt:  [+3128.1% +3142.7% +3156.6%]
                        Performance has improved.

FFT twiddles generation/natural inversed                                                                            
                        time:   [19.369 ms 19.566 ms 19.702 ms]
                        thrpt:  [53.223 Melem/s 53.593 Melem/s 54.136 Melem/s]
                 change:
                        time:   [-97.226% -97.208% -97.189%] (p = 0.00 < 0.05)
                        thrpt:  [+3457.6% +3481.0% +3504.4%]
                        Performance has improved.

FFT twiddles generation/bit-reversed                                                                            
                        time:   [29.767 ms 29.791 ms 29.826 ms]
                        thrpt:  [35.157 Melem/s 35.198 Melem/s 35.226 Melem/s]
                 change:
                        time:   [-95.303% -95.254% -95.213%] (p = 0.00 < 0.05)
                        thrpt:  [+1988.8% +2007.2% +2028.9%]
                        Performance has improved.

FFT twiddles generation/bit-reversed inversed                                                                            
                        time:   [29.821 ms 29.909 ms 30.006 ms]
                        thrpt:  [34.945 Melem/s 35.059 Melem/s 35.163 Melem/s]
                 change:
                        time:   [-95.698% -95.684% -95.668%] (p = 0.00 < 0.05)
                        thrpt:  [+2208.3% +2216.9% +2224.6%]
                        Performance has improved.

Bit-reverse permutation/Sequential                                                                            
                        time:   [24.589 ms 24.667 ms 24.824 ms]
                        thrpt:  [84.480 Melem/s 85.017 Melem/s 85.289 Melem/s]
                 change:
                        time:   [-9.3776% -6.9110% -4.4375%] (p = 0.00 < 0.05)
                        thrpt:  [+4.6435% +7.4240% +10.348%]
                        Performance has improved.

Polynomial evaluation/Sequential FFT                                                                           
                        time:   [718.67 ms 721.94 ms 724.82 ms]
                        thrpt:  [2.8933 Melem/s 2.9049 Melem/s 2.9181 Melem/s]
                 change:
                        time:   [-52.667% -52.025% -51.465%] (p = 0.00 < 0.05)
                        thrpt:  [+106.04% +108.44% +111.27%]
                        Performance has improved.

Polynomial interpolation/Sequential FFT                                                                           
                        time:   [755.05 ms 756.12 ms 757.24 ms]
                        thrpt:  [2.7695 Melem/s 2.7736 Melem/s 2.7775 Melem/s]
                 change:
                        time:   [-52.984% -52.796% -52.608%] (p = 0.00 < 0.05)
                        thrpt:  [+111.01% +111.85% +112.69%]
                        Performance has improved.
```

### CODE REVIEW

File: fft/src/roots_of_unity.rs [hunk 1]
```diff
@@ -17,16 +17,28 @@ pub fn get_powers_of_primitive_root<F: IsFFTField>(
...
-        RootsConfig::BitReverse | RootsConfig::BitReverseInversed => {
-            root.pow(reverse_index(&i, count as u64))
+    if matches!(
+        config,
+        RootsConfig::BitReverse | RootsConfig::BitReverseInversed
+    ) {
+        let mut reversed = Vec::with_capacity(count);
+        for i in 0..count {
+            reversed.push(results[reverse_index(&i, count as u64)].clone());
         }
```
Comment (MegaRedHand) [inline fft/src/roots_of_unity.rs:38]: I'm curious if using `extend` would improve performance:
```suggestion
        let iter = (0..count).into_iter().map(|i| results[reverse_index(&i, count as u64)].clone());
        reversed.extend(iter);
```

**Review by MegaRedHand — APPROVED** [review]
🚀 🚀

File: fft/src/roots_of_unity.rs (hunk 1 above)
Comment (Oppen) [reply fft/src/roots_of_unity.rs:38]: I could try that.

Comment (Oppen) [reply fft/src/roots_of_unity.rs]: It did produce a significant improvement. 10% for non-inversed and 7% for inversed.

**Review by xqft — APPROVED** [review]
wonderful

**Review by pablodeymo — APPROVED** [review]
LGTM

**Pull request merged**